designs/2022-async-parse/README.md

```diff
@@ -0,0 +1,238 @@
+
+#### Details
+
+TODO: as these are internal tools, I left out the details from the first version of this RFC.
+
+## Documentation
+
+* [Working with Custom Parsers](https://github.com/eslint/eslint/blob/main/docs/developer-guide/working-with-custom-parsers.md) should be updated with info that `parse` / `parseForESLint` can return a Promise, and an example for `async parseForESLint()`.
+* [Node.js API - Linter](https://github.com/eslint/eslint/blob/main/docs/developer-guide/nodejs-api.md#linter) should be updated with the new return types for `verify` and `verifyAndFix`.
+* [Node.js API - RuleTester](https://github.com/eslint/eslint/blob/main/docs/developer-guide/nodejs-api.md#ruletester) should be updated with new requirements for its use.
+* We should announce the breaking changes in a "What's coming in ESLint v9.0.0" blog post.
+
+## Drawbacks
+
+* Added complexity. While the code stays pretty much the same thanks to the async-await syntax, asynchronous execution is inherently more complex to reason about and brings additional concerns such as concurrency issues.
+* Performance with sync parsers. Converting a sync flow to an async flow with the exact same steps should add some time to the execution, though I don't see any differences when running `npm run perf` on the [async-parse](https://github.com/mdjermanovic/eslint/tree/async-parse) branch.
+* This requires breaking changes in the `Linter` API and new requirements for the use of `RuleTester`.
+
+## Backwards Compatibility Analysis
+
+### End Users
+
+There should be no breaking changes for end users.
+
+### API
+
+Breaking changes in the public API:
+
+* `Linter#verify` will always return `Promise<LintMessage[]>` instead of `LintMessage[]`.
+* `Linter#verifyAndFix()` will always return `Promise<{fixed:boolean,messages:LintMessage[],output:string}>` instead of `{fixed:boolean,messages:LintMessage[],output:string}`.
+
+We provide the `Linter` class mostly for the use in browsers. Projects using this class will have to update their code for the new async API.
+
+ESLint integrations are expected to use the `ESLint` class. There will be no changes for them.
+
+### Plugins
+
+Breaking changes for plugins are only related to `RuleTester`:
+
+* `RuleTester` will require a testing framework or custom `it` handler.
+* `it` must support returning a Promise from passed callbacks. It should await for the Promise to be settled and treat the test as "passing" only if the Promise is fulfilled. If the Promise is not settled within a certain timeframe or is rejected, the test should be treated as "failing". Mocha and Jest already support this. 
+
+Projects that use `RuleTester` with Mocha or Jest wouldn't be required to change anything.
+
+Projects that run tests directly or use a testing framework that doesn't support returning a Promise from callbacks will have to switch to a framework like Mocha or Jest.
+
+## Alternatives
+
+### Alternatives to `async parse()` / `async parseForESLint()`
+
```

Comment:
At the moment I have function configs disabled in flat config due to the complexity of getting it working.

Reply:
Another option: we could actually do the async loading inside of the ESLint class as part of preprocessing and then stick the loaded module directly into the config that is passed to Linter.

Reply:
> Another option: we could actually do the async loading inside of the ESLint class as part of preprocessing and then stick the loaded module directly into the config that is passed to Linter.

I've added this alternative and another one with top-level await.